### ISSUE TITLE
Probing Leaf..

### ISSUE THREAD

**Issue by jonysy** [lychee-eng/leaf#1]
#### Medium posts:

* [Rust Guide to CHANGELOG.md](https://medium.com/@autumn_eng/guide-to-changelog-md-in-rust-6eb349808fa4)
* [About Rust’s Machine Learning Community](https://medium.com/@autumn_eng/about-rust-s-machine-learning-community-4cda5ec8a790)
* [Increasing Open-Source Engagement w/ Structural Communication Guidelines for Code Documentation](https://medium.com/@autumn_eng/increasing-open-source-engagement-with-structural-communication-guidelines-for-code-documentation-e72533de8e45)
* [Tensorflow wins](https://medium.com/@mjhirn/tensorflow-wins-89b78b29aafb)

#### On automatic differentiation (AD) and optimizations:

gitter:

> **Michael Hirn**: We are currently working on the backend abstraction and parallelisation for CPU and GPU support.
> We were looking into auto diff a bit already.
> I will link quickly to the reddit comments about auto diff on leaf: [discussion](https://www.reddit.com/r/rust/comments/3s9z2g/leaf_machine_intelligence_framework_in_rust/)
> auto diff would be nice, but it comes with computational cost, which we can avoid in plain ML scenarios. So not a super high prio at the moment.
> **Dominique Luna**: I was thinking something along the lines of [repo](https://github.com/HIPS/autograd) and [repo](https://github.com/twitter/torch-autograd) rather than Theano. I think Tensorflow is somewhere in between.
> As for performance, you're right it comes at a cost but according to this [post](https://blog.twitter.com/2015/autograd-for-torch)  it's typically only 2x and they still have optimizations they can make. Also, if you're going to consider the computer time you also should consider the developer's time should is worth more and auto diff helps tremendously.
> ....
> **MH**: ..will receive a hard refactor to support the backend abstraction (mem synchronization) and then of course the new backend crate to support executing the computations across platforms and devices..

**Comment by jonysy** [comment]
#### Medium posts:

* [Rust Guide to CHANGELOG.md](https://medium.com/@autumn_eng/guide-to-changelog-md-in-rust-6eb349808fa4)
* [About Rust’s Machine Learning Community](https://medium.com/@autumn_eng/about-rust-s-machine-learning-community-4cda5ec8a790)
* [Increasing Open-Source Engagement w/ Structural Communication Guidelines for Code Documentation](https://medium.com/@autumn_eng/increasing-open-source-engagement-with-structural-communication-guidelines-for-code-documentation-e72533de8e45)
* [Tensorflow wins](https://medium.com/@mjhirn/tensorflow-wins-89b78b29aafb)

**Comment by jonysy** [comment]
#### On automatic differentiation (AD) and optimizations:

gitter:

> **Michael Hirn**: We are currently working on the backend abstraction and parallelisation for CPU and GPU support.
> We were looking into auto diff a bit already.
> I will link quickly to the reddit comments about auto diff on leaf: [discussion](https://www.reddit.com/r/rust/comments/3s9z2g/leaf_machine_intelligence_framework_in_rust/)
> auto diff would be nice, but it comes with computational cost, which we can avoid in plain ML scenarios. So not a super high prio at the moment.
> **Dominique Luna**: I was thinking something along the lines of [repo](https://github.com/HIPS/autograd) and [repo](https://github.com/twitter/torch-autograd) rather than Theano. I think Tensorflow is somewhere in between.
> As for performance, you're right it comes at a cost but according to this [post](https://blog.twitter.com/2015/autograd-for-torch)  it's typically only 2x and they still have optimizations they can make. Also, if you're going to consider the computer time you also should consider the developer's time should is worth more and auto diff helps tremendously.
> ....
> **MH**: ..will receive a hard refactor to support the backend abstraction (mem synchronization) and then of course the new backend crate to support executing the computations across platforms and devices..

**Comment by jonysy** [comment]
#### Declarative (Tensorflow, Theano, etc.) vs Imperative (Leaf, Torch, etc.) approaches

[hn](https://news.ycombinator.com/item?id=11563846):

> I think CNTK and Tensorflow and Theano, too, have a declarative approach, representing the computation via a computational graph. Which in my opinion is beneficial for research. But for a hacker or a software developer who wants to build an application, this creates an unnecessarily steep learning curve and feels unintuitive. (I have the feeling, that this an important reason why Keras, Lasagne and co. exists)
>
> Leaf takes an imperative approach and explores an easier API (only Layers [Functions](http://autumnai.com/leaf/book/deep-learning-glossary.html#Layer) and Solvers (Optimizer Algorithms)), reusability through modularity and abstractions that keep the implementation and concepts to a minimum or rather abstractions that feel as familiar to a hacker as possible.
>
> For future versions e.g., we want to explore what is practically possible with auto-differentiation via dual numbers and differentiable programming. 

Solvers:

> We wanted to make clear that we are not "just" doing deep learning with Leaf. A lot of other ML algorithms that don't fall under the deep learning label can benefit of the tools provided (e.g. solvers) and can be implemented as layers. While not too popular, there are papers showing this being done with e.g. a [SVM](http://deeplearning.net/wp-content/uploads/2013/03/dlsvm.pdf).

[hn](https://news.ycombinator.com/item?id=10541568):
> **rryan**: Great work -- very exciting to see a Rust ML framework!
>
> Dist-Belief (TensorFlow's predecessor) was oriented around layers (as Leaf is) but TensorFlow (like Theano) is graph/flow-based rather than layer-based. I think in particular the layer-based approach makes it awkward to represent more complicated models since not every model can be decomposed into backward/forward passes.
>
> Do you worry that by adopting a layer-based paradigm you are boxing yourself in? A graph-based approach does seem more flexible (and distributable across machines).
>
> **M. Goisser**: > I am very optimistic concerning our plans to make Leaf work in a distributed fashion. It's not quite finalized yet, but the rough outline consists of abstracting other parts of the Network away as a special type of layer and using capnproto + zeromq for transportation. Sadly I haven't come across a good implementation of model parallelism to learn from, so that might become a bit tricky.
>
> I think there is still a lot of room for improvement for layer-based networks, but even if the graph-based approach will become more prelevant I think the Rust trait system gives us great tools to migrate and reuse a lot from our layer based approach.
>
> **Djvulee**: I just curious about is there any good Actor model library in Rust, just like Akka for Scala/Java, so there is a better model for transportation.

**Comment by jonysy** [comment]
#### Non-sequential containers

gitter:

> **M. Goisser**: You can get pretty far in composing a directed graph with sequential layers alone, and in my opinion that should be the main Leaf interface. There are however a few utility layers that are missing and would allow for more options.
>
> For example a Split layer.
>
> There is no Function layer, and I think that that would only really work on the CPU.
>
> I feel like you are looking for a more symbolic aproach than what Leaf offers?

**Comment by jonysy** [comment]
#### Alexander Morozov comments

gitter: 

> About roadmap I'd suggest 1) merge most current PRs, they are mostly approved, just stalled, 2) cleanup Leaf code, there is a lot of boilerplate, е. g. I think Arc<RwLocks<>> are not actually required, but I'm not completely sure, 3) fix GPU underutilization issues: published benchmarks AFAIK test just forward/backward passes without I/o and solver parts. When they are added, Leaf is not so fast - nvidia-smi showed that GPU utilization is about 40% on my low-mid range card. I've submitted several PRs that reduce CUDA i/o and increase performance 10-30%, and I'm sure a lot more can be done.
>
> About (3) -- I've got some unbased speculations that "imperative-style" approach will eventually limit possible performance, as it would be difficult to submit several commands to GPU at once. It's certainly possible, but it will increase code complexity, while Leaf goal was to keep code as simple and hackable as possible. @/MichaelHirn didn't agree with me on this and he definitely understands NNs and GPUs better, so I think best way to find out is to take time to profile a complete NN with solver and find out what bottlenecks currently are there and if they can be removed for wide enough range of networks.

**Comment by jonysy** [comment]
#### Supporting the creation of dynamic computation graphs (DCG) or static computation graphs (SCG)

Todo..
